### PR TITLE
Speed up torch.unique_consecutive()

### LINKED ISSUE #62690: torch.unique_consecutive() is very slow when dim is specified even with 1-d tensors
## Issue description

torch.unique_consecutive(dim=0) is very slow with 1-d tensors (and also with 2-d tensors). Please see the following. 

## Code example

torch.manual_seed(0)
t = torch.randint(500, (10000000, ))
t = torch.sort(t)[0]

start = time.time()
uniques, counts = torch.unique_consecutive(t, dim=0, return_counts=True)
end = time.time()
print("torch.unique_consecutive(dim=0) time:", end - start)

start = time.time()
uniques, counts = torch.unique_consecutive(t, return_counts=True)
end = time.time()
print("torch.unique_consecutive() time:", end - start)

## Code output

torch.unique_consecutive(dim=0) time: 27.382863759994507
torch.unique_consecutive() time: 0.011050701141357422

## System Info
Collecting environment information...
PyTorch version: 1.9.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.8.8 (default, Apr 13 2021, 19:58:26)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-1055-azure-x86_64-with-glibc2.10
Is CUDA available: False
CUDA runtime version: 11.1.74
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] numpydoc==1.1.0
[pip3] torch==1.9.0+cu111
[pip3] torch-geometric==1.7.2
[pip3] torch-scatter==2.0.7
[pip3] torch-sparse==0.6.10
[pip3] torch-tb-profiler==0.2.1
[pip3] torchaudio==0.9.0
[pip3] torchvision==0.10.0+cu111
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py38h27cfd23_1  
[conda] mkl_fft                   1.3.0            py38h42c9631_2  
[conda] mkl_random                1.2.1            py38ha9443f7_2  
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     1.9.0+cu111              pypi_0    pypi
[conda] torch-geometric           1.7.2                    pypi_0    pypi
[conda] torch-scatter             2.0.7                    pypi_0    pypi
[conda] torch-sparse              0.6.10                   pypi_0    pypi
[conda] torch-tb-profiler         0.2.1                    pypi_0    pypi
[conda] torchaudio                0.9.0                    pypi_0    pypi
[conda] torchvision               0.10.0+cu111             pypi_0    pypi

cc @VitalyFedyunin @ngimel @heitorschueroff

### PR DESCRIPTION
**Pull request by oToToT**
Fixes #62690

## Summary
Like the way `unique_consecutive_cpu_template` implemented, this PR reimplements `_unique_dim_cpu_impl` to get better performance.
Also, because the overhead of `unique_dim_consecutive_cpu` is quite large, directly call `unique_consecutive_cpu_template` when we know the given input is a 1d-array.

### Script
```python
import torch
import time

torch.manual_seed(0)
t = torch.randint(500, (10000000, ))
t = torch.sort(t)[0]

start = time.time()
uniques, inverse, counts = torch.unique_consecutive(t, dim=0, return_inverse=True, return_counts=True)
end = time.time()
print("torch.unique_consecutive(dim=0) time:", end - start)

start = time.time()
uniques2, inverse2, counts2 = torch.unique_consecutive(t, return_inverse=True, return_counts=True)
end = time.time()
print("torch.unique_consecutive() time:", end - start)


t = torch.randint(500, (10000000, 2))
t = torch.sort(t)[0]

start = time.time()
uniques, inverse, counts = torch.unique_consecutive(t, dim=0, return_inverse=True, return_counts=True)
end = time.time()
print("torch.unique_consecutive(dim=0) time:", end - start)

start = time.time()
uniques, inverse, counts = torch.unique_consecutive(t, dim=1, return_inverse=True, return_counts=True)
end = time.time()
print("torch.unique_consecutive(dim=1) time:", end - start)
```

### Before
```
torch.unique_consecutive(dim=0) time: 78.64345622062683
torch.unique_consecutive() time: 0.029544353485107422
torch.unique_consecutive(dim=0) time: 91.49796152114868
torch.unique_consecutive(dim=1) time: 0.30872368812561035
```

### After
```
torch.unique_consecutive(dim=0) time: 0.08256125450134277
torch.unique_consecutive() time: 0.08162403106689453
torch.unique_consecutive(dim=0) time: 35.58408498764038
torch.unique_consecutive(dim=1) time: 1.6258199214935303
```

## System Information
```
Collecting environment information...
PyTorch version: 1.10.0a0+git7f1932e
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun  2 2021, 10:49:15)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.11.0-34-generic-x86_64-with-glibc2.29
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.10.0a0+gitbe09195
[conda] Could not collect
```

### CODE REVIEW

File: aten/src/ATen/native/Unique.cpp [hunk 1]
```diff
@@ -135,24 +135,23 @@ ForwardIt _unique_dim_cpu_impl(ForwardIt first, ForwardIt last,
...
       return last;
     }
-    // save to calculate distance to iterators
-    ForwardIt begin = first;
 
-    // set first inverse index and count
-    inverse_indices_vec[indices[0]] = 0;
-    counts[0] += 1;
+    int64_t *indices_data = indices.data();
+    int64_t *inverse_data = inverse_indices_vec.data_ptr<int64_t>();
```
Comment (ngimel) [inline aten/src/ATen/native/Unique.cpp:140]: Can you please add a TORCH_INTERNAL_ASSERT that inverse_indices_vec and counts are contiguous? They currently are at the only callsite, but this function should explicitly check its assumptions.

Comment (oToToT) [reply aten/src/ATen/native/Unique.cpp:145]: Sure, added.

File: aten/src/ATen/native/Unique.cpp [hunk 2]
```diff
@@ -275,7 +279,7 @@ unique_dim_consecutive_cpu(const Tensor& self, const int64_t dim, const bool ret
 
 std::tuple<Tensor, Tensor, Tensor>
 unique_consecutive_cpu(const Tensor& self, const bool return_inverse, const bool return_counts, c10::optional<int64_t> dim) {
-  if (!dim.has_value()) {
+  if (!dim.has_value() || (dim.value() == 0 && self.sizes().size() == 1)) {
```
Comment (oToToT) [inline aten/src/ATen/native/Unique.cpp:282]: Without this check, the result would be

```
torch.unique_consecutive(dim=0) time: 16.88133478164673
torch.unique_consecutive() time: 0.024564027786254883
torch.unique_consecutive(dim=0) time: 28.210450410842896
torch.unique_consecutive(dim=1) time: 0.30884695053100586
```

As you could see, this check slightly slows down `torch.unique_consecutive(dim=1)`, but it could also speed up `torch.unique_consecutive(dim=0)` very much.
Hence, I still leave it here.

Comment (ngimel) [reply aten/src/ATen/native/Unique.cpp:282]: How can this check affect `dim=1` case? With or without this check `dim=1` case should go to `unique_dim_consecutive_cpu` in line 287. 5x slowdown (from 0.3 to 1.6) seems significant, why is your algo slower than the previous one for `dim=1`?

Comment (oToToT) [reply aten/src/ATen/native/Unique.cpp:282]: Maybe some CPU branching issue?

If I tried loop the last case 5 times, the result become
```
torch.unique_consecutive(dim=0) time: 0.06868529319763184
torch.unique_consecutive() time: 0.06931805610656738
torch.unique_consecutive(dim=0) time: 26.346323251724243
torch.unique_consecutive(dim=1) time: 1.2163631916046143
torch.unique_consecutive(dim=1) time: 0.2443101406097412
torch.unique_consecutive(dim=1) time: 0.24067211151123047
torch.unique_consecutive(dim=1) time: 0.25301051139831543
torch.unique_consecutive(dim=1) time: 0.24738121032714844
```

I will try `torch.utils.benchmark` or something else to get a more thorough analysis if possible.

Comment (ngimel) [reply aten/src/ATen/native/Unique.cpp]: Ah ok, it seems like a warm-up issue then, but please post torch.utils.benchmark timings here, so that we have a record.

Comment (oToToT) [reply aten/src/ATen/native/Unique.cpp]: The script I test [test_1d.py](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-test_1d-py), [test_2d.py](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-test_2d-py).

Result on be091950: [1d result](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-1d_result_be091950-txt), [2d result](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-2d_result_be091950-txt)

Result on 89e66488 w/ `(dim.value() == 0 && self.sizes().size() == 1)`: [1d result](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-1d_result_89e66488-txt), [2d result](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-2d_result_89e66488-txt)

Result on 89e66488 w/o `(dim.value() == 0 && self.sizes().size() == 1)`: [1d result](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-1d_result_89e66488_no_check-txt), [2d result](https://gist.github.com/oToToT/ba1a8951d55628354a03ac18d3871d2c#file-2d_result_89e66488_no_check-txt)

Summary:
For 1d case, >100x more faster on `dim=0` case with `(dim.value() == 0 && self.sizes().size() == 1)`.
For 2d case, there is no significant impact with or without `(dim.value() == 0 && self.sizes().size() == 1)`.
In general, the new implementation (89e66488) is 3x more faster than the original implementation (be091950).

**Review by ngimel — APPROVED** [review]
Thanks for the improvement!
